engine/src/cmd/collision.cpp

```diff
@@ -268,12 +262,78 @@ void Collision::collide( Unit* unit1,
+            );
+    QVector impulse = impulse_magnitude * normal1;
+
+    QVector v1_new = unit1->GetVelocity() - impulse * mass1_inverse;
+    QVector v2_new = unit2->GetVelocity() + impulse * mass2_inverse; // difference in sign is intentional and important
+    QVector w1_new = unit1->GetAngularVelocity() - impulse_magnitude * I1_inverse * (location1_local.Cross(normal1));
+    QVector w2_new = unit2->GetAngularVelocity() + impulse_magnitude * I2_inverse * (location2_local.Cross(normal1)); // again, repeated use of normal1 and lack of use of normal2 is intentional; difference in sign is intentional and important
+
+    /* 
+    * Can factor the following out into testing code later
+    * BEGIN TESTING CODE
+    */
+    validateCollision(relative_velocity, normal1, location1_local, location2_local, v1_new, v2_new, w1_new, w2_new);
+    /*
+    * END TESTING CODE
+    */
+    
+    // Kinetic energy including rotational, from https://en.wikipedia.org/wiki/Moment_of_inertia#Kinetic_energy_2 
+    double kinetic_energy_system_initial_linear = 0.5 * mass1 * unit1->GetVelocity().MagnitudeSquared() + 0.5 * mass2 * unit2->GetVelocity().MagnitudeSquared();
+    double kinetic_energy_system_initial_angular = 0.5 * unit1->GetAngularVelocity().Dot(I1 * unit1->GetAngularVelocity()) + 0.5 * unit2->GetAngularVelocity().Dot(I2 * unit2->GetAngularVelocity());
+    
+    // compute changes in kinetic energy for linear and rotational
+    double kinetic_energy_system_linear = 0.5 * mass1 * v1_new.MagnitudeSquared() + 0.5 * mass2 * v2_new.MagnitudeSquared();
+    double kinetic_energy_system_angular = (0.5 * w1_new.Dot(I1 * w1_new)) + ( 0.5 * w2_new.Dot(I2 * w2_new));
+    double delta_kinetic_energy_linear = kinetic_energy_system_initial_linear - kinetic_energy_system_linear;
+    double delta_kinetic_energy_angular = kinetic_energy_system_initial_angular - kinetic_energy_system_angular;
+    
+    //we have an impulse [ Force * Time ] and our physics interfaces operate on forces and torques [Force * Meters]
+    QVector force_on_1 = -1 * impulse * (1.0f / (unit1->sim_atom_multiplier * SIMULATION_ATOM)); // divide impulse by time over which it will be applied to derive force -- allow different units to have different physics fidelity
+    QVector force_on_2 = impulse * (1.0f / (unit2->sim_atom_multiplier * SIMULATION_ATOM)); // divide impulse by time over which it will be applied to derive force 
+
+    collision1.adjustInterpenetration(v1_new, w1_new, normal1); // Call this BEFORE applying force
+    collision1.applyForce(force_on_1, location1_local); // handles both force-on-center-of-mass and torque
+    collision2.adjustInterpenetration(v2_new, w2_new, normal1); // uses normal1, not normal2
+    collision2.applyForce(force_on_2, location2_local);
+
+    collision1.dealDamage(collision2, delta_kinetic_energy_linear, delta_kinetic_energy_angular);
+    collision2.dealDamage(collision1, delta_kinetic_energy_linear, delta_kinetic_energy_angular);
```

Comment:
Wow. My head is spinning right about now. Nice job, @ministerofinformation

Reply:
If your head is spinning, are there more comments I could add to make it more self-contained without dumping the whole Wikipedia article contents in (the figure is actually quite useful for building an intuition of what's going on)? I'd like this to be as clearly documented as possible, so as to separate the math of the collision from future renovations on the moment of inertia implementation and various getters/non-local invariants we want to adjust/enforce (applying Newtonian physics to zero-mass objects tends to end poorly in more than the collision code, etc.)

Reply:
No, it's fine, thanks. Just a lot of code to read through. :-)

Reply:
Just standard vector arithmetic @stephengtuggy